pageserver/src/layered_repository.rs

```diff
@@ -829,12 +849,7 @@ impl Timeline for LayeredTimeline {
     /// NOTE: this updates last_valid_lsn as well.
     ///
     fn advance_last_record_lsn(&self, lsn: Lsn) {
-        // Can't move backwards.
-        let old = self.last_record_lsn.fetch_max(lsn);
-        assert!(old <= lsn);
-
-        // Use old value of last_record_lsn as prev_record_lsn
-        self.prev_record_lsn.fetch_max(old);
+        self.record_lsn.write().unwrap().advance(lsn);
 
         // Also advance last_valid_lsn
         let old = self.last_valid_lsn.advance(lsn);
```

Comment:
To satisfy the invariant `last_valid_lsn >= last_record_lsn`, I think we need the following orderings:

Loads (as you suggested):
1. `record_lsn`
2. `valid_lsn`

Advances:
1. `valid_lsn`
2. `record_lsn`

We should re-order the advances here (in `advance_last_record_lsn`) to match `init_valid_lsn`